src/System.Net.Http/src/System/Net/Http/DiagnosticsHandler.cs

```diff
@@ -0,0 +1,176 @@
+            //let inner handler decide what to do with the it
+            if (request == null || cancellationToken.IsCancellationRequested)
+            {
+                return await base.SendAsync(request, cancellationToken);
+            }
+
+            Activity requestActivity = null;
+
+            //Check if DiagnosticSource is enabled for this RequestUri: 
+            //Client code may subscribe to Http Diagnostic Listener with
+            // - no predicate: in this case it will receive ALL events for all outgoing requests which are
+            //      System.Net.Http.Request
+            //      System.Net.Http.Response
+            //      System.Net.Http.Activity.Start
+            //      System.Net.Http.Activity.Stop
+            // - Predicate: Func<string, bool> to filter event names and URIs:
+            //      If user does not want to receive any System.Net.Http.Activity.* events, 
+            //        he may subscribe to Request, Response events only,
+            //      If user wants to filter particular URIs, he may write predicate to achieve it
+            // Also check for Activity.Current: if it's null (subject to change to Activity.IsEnabled) 
+            // this request is not sampled and we don't inject headers
+            if (Activity.Current != null &&
+                s_diagnosticListener.IsEnabled(request.RequestUri.ToString()))
+            {
+                // create a new activity for the outgoing Http request
+                requestActivity = new Activity(HttpHandlerLoggingStrings.HttpActivityWriteName)
+                    .SetStartTime(DateTimeStopwatch.GetTime());
+                
+                //Start it and notify subscribers
+                s_diagnosticListener.StartActivity(requestActivity, new {Request = request});
+
+                //Inject correlation headers
+                request.Headers.Add(RequestIdHeaderName, requestActivity.Id);
+                List<string> baggage = FormatBaggageHeader(requestActivity.Baggage);
+                if (baggage.Count != 0)
+                {
+                    request.Headers.Add(CorrelationContextHeaderName, baggage);
+                }
+            }
+
+            //notify subscribers listening to Request, Response events
+            //System.Net.Http.Request and Response events may be eventually deprecated 
+            //for now let's keep those events, so existing users will not be affected
+            Guid loggingRequestId = LogHttpRequestCore(request);
+
+            HttpResponseMessage response = null;
+            Exception exception = null;
+            try
+            {
+                response = await base.SendAsync(request, cancellationToken);
+                //previously Response event was only fired if request was sucessful, 
+                //this behavior does not change
+                LogHttpResponseCore(response, loggingRequestId);
```

Comment:
I was afraid to fix it, because there could be an assumption that Response event means that request was successful.
I also think that we should deprecate Request/Response events eventually and it make no sense to fix this behavior